File: 323.number-of-connected-components-in-an-undirected-graph.cpp
/*
 * @lc app=leetcode id=323 lang=cpp
 *
 * [323] Number of Connected Components in an Undirected Graph
 */

// @lc code=start
class Solution {
public:
    int countComponents(int n, vector<vector<int>>& edges) {
        vector<int> reps(n);
        iota(reps.begin(), reps.end(), 0);

        for(auto &edge: edges) {
            combine(reps, edge[0], edge[1]);
        }

        unordered_set<int> ans;
        for(int i=0; i<n; ++i) {
            ans.insert(find(reps, i));
        }

        return ans.size();
    }

    int find(vector<int>& reps, int x) {
        if(reps[x] == x) return x;
        return reps[x] = find(reps, reps[x]);
    }

    void combine(vector<int>& reps, int x, int y) {
        reps[find(reps, x)] = find(reps, y);
    }
};
// @lc code=end

File: 323.number-of-connected-components-in-an-undirected-graph.py
#
# @lc app=leetcode id=323 lang=python3
#
# [323] Number of Connected Components in an Undirected Graph
#

# @lc code=start
class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        parent = list(range(n))

        def find(x: int) -> int:
            if parent[x] != x:
                parent[x] = find(parent[x])
            return parent[x]
        
        def union(x: int, y: int) -> None:
            px, py = find(x), find(y)
            if px < py:
                parent[py] = px
            else:
                parent[px] = py

        for x, y in edges:
            union(x, y)
        return len({find(x) for x in parent})

# @lc code=end

